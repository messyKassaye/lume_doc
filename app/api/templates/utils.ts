import uuid from 'node-uuid';
import { PropertySchema } from 'shared/types/commonTypes';
import settings from 'api/settings/settings';
import { ThesaurusValueSchema } from 'shared/types/thesaurusType';

const generateNewSafeName = (label: string) =>
  label
    .trim()
    .replace(/[#|\\|/|*|?|"|<|>|=|||\s|:|.]/gi, '_')
    .replace(/^[_|\-|+|$]/, '')
    .toLowerCase();

export const safeName = (label: string, newNameGeneration: boolean = false) => {
  if (newNameGeneration) {
    return generateNewSafeName(label);
  }
  return label
    .trim()
    .replace(/[^a-z0-9]/gi, '_')
    .toLowerCase();
};

const generateName = (property: PropertySchema, newNameGeneration: boolean) => {
  const name = property.label ? safeName(property.label, newNameGeneration) : property.name;
  return property.type === 'geolocation' || property.type === 'nested'
    ? `${name}_${property.type}`
    : name;
};

const generateNames = (properties: PropertySchema[], newNameGeneration: boolean) =>
  properties.map(property => ({
    ...property,
    name: generateName(property, newNameGeneration),
  }));

export function generateIds(properties: PropertySchema[] = []) {
  return properties.map(property => ({
    ...property,
    id: property.id || uuid.v4(),
  }));
}

export const generateNamesAndIds = async (properties: PropertySchema[] = []) => {
  const { newNameGeneration = false } = await settings.get();
  return generateIds(generateNames(properties, newNameGeneration));
};

export interface PropertyOrThesaurusSchema
  extends Partial<PropertySchema>,
    Partial<ThesaurusValueSchema> {}

const flattenProperties = (properties: PropertyOrThesaurusSchema[]) =>
  properties.reduce<PropertyOrThesaurusSchema[]>((flatProps, p) => {
    if (p.values) {
      return [...flatProps, ...p.values];
    }

    return [...flatProps, p];
  }, []);

export function getUpdatedNames(
  oldProperties: PropertyOrThesaurusSchema[] = [],
  newProperties: PropertyOrThesaurusSchema[],
  prop: 'name' | 'label' = 'name',
  outKey: 'name' | 'label' = prop
) {
  const propertiesWithNewName: { [k: string]: string | undefined } = {};

  flattenProperties(oldProperties).forEach(property => {
    const newProperty = flattenProperties(newProperties).find(p => p.id === property.id);
    if (newProperty && newProperty[prop] !== property[prop]) {
      const key = property[outKey];
      if (key) {
        propertiesWithNewName[key] = newProperty[prop];
      }
    }
  });

  return propertiesWithNewName;
}

const notIncludedIn = (propertyCollection: PropertyOrThesaurusSchema[]) => (
  property: PropertyOrThesaurusSchema
) => !propertyCollection.find(p => p.id === property.id);

export function getDeletedProperties(
  oldProperties: PropertyOrThesaurusSchema[] = [],
  newProperties: PropertyOrThesaurusSchema[],
  prop: 'name' | 'label' = 'name'
) {
  return flattenProperties(oldProperties)
    .filter(notIncludedIn(flattenProperties(newProperties)))
    .map(property => property[prop]);
}
