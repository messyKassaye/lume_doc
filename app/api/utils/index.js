import validateRequest from './validateRequest';
import createError from './Error';
import handleError from './handleError';
import { parseQuery } from './parseQueryMiddleware';

const validation = {
  validateRequest,
};

export { validation, createError, handleError, parseQuery };
