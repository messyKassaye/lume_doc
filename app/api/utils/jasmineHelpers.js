/** @format */

const assessStatus = routeResult => {
  let status;
  const req = {};
  const next = () => {};
  const res = {
    status: code => {
      status = code;
    },
    json: () => {},
  };

  routeResult.middlewares[0](req, res, next);

  return status;
};

const assessAuthorized = (routeResult, roles = []) => {
  const nextCalledFor = [];

  roles.forEach(role => {
    const req = {
      user: { role },
      get: key => (key === 'X-Requested-With' ? 'XMLHttpRequest' : ''),
    };
    const res = { status: () => {}, json: () => {} };
    const next = () => {
      nextCalledFor.push(role);
    };
    routeResult.middlewares[0](req, res, next);
  });

  return roles.join(',') === nextCalledFor.join(',');
};

const conformValidationErrors = (routeCanFail, expected) =>
  routeCanFail
    ? `Route is not correctly authorized for ['${expected.join("', '")}']`
    : 'Route is not authorized ! (Auth middleware should be the first one)';

const conformMessage = (result, routeCanFail, expected) =>
  result.pass ? 'route is authorized' : conformValidationErrors(routeCanFail, expected);

export function catchErrors(done) {
  return error => {
    if (error instanceof Error) {
      return done.fail(error);
    }
    return done.fail(JSON.stringify(error));
  };
}

const matchers = {
  toNeedAuthorization() {
    return {
      compare(routeResult, expected) {
        const routeCanFail = assessStatus(routeResult) === 401;
        const routeValidatesExpected = expected ? assessAuthorized(routeResult, expected) : true;
        const result = { pass: routeCanFail && routeValidatesExpected };
        result.message = () => conformMessage(result, routeCanFail, expected);
        return result;
      },
    };
  },
  containItems() {
    return {
      compare(actual, expected) {
        const result = {};
        const missingItems = expected.filter(item => !actual.includes(item));
        result.pass = !missingItems.length && actual.length === expected.length;
        if (result.pass) {
          result.message = 'Collections are equal';
        } else {
          result.message = `Expected: [${expected}] but got: [${actual}]`;
        }
        return result;
      },
    };
  },
  toContainEqual: jasmine.matchers ? jasmine.matchers.toContain : () => {},
};

beforeEach(() => {
  jasmine.addMatchers(matchers);
});
