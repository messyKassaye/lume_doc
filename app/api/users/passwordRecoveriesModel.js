import mongoose from 'mongoose';

import { instanceModel } from 'api/odm';

const schema = new mongoose.Schema({
  key: String,
  user: { type: mongoose.Schema.Types.ObjectId, ref: 'users' },
});

export default instanceModel('passwordrecoveries', schema);
