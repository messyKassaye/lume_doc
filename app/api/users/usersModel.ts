import mongoose from 'mongoose';

import { instanceModel } from 'api/odm';

export interface User {
  _id?: any;
  username?: string;
  password?: string;
  email?: string;
  role?: string;
  failedLogins?: number;
  accountLocked?: boolean;
  accountUnlockCode?: string;
  using2fa?: boolean;
  secret?: string;
}

const userSchema = new mongoose.Schema({
  username: { type: String, unique: true, required: true },
  password: { type: String, select: false, required: true },
  email: { type: String, unique: true, required: true },
  role: { type: String, unique: false, required: true },
  failedLogins: { type: Number, required: false, select: false },
  accountLocked: { type: Boolean, select: false },
  accountUnlockCode: { type: String, select: false },
  using2fa: { type: Boolean, default: false },
  secret: { type: String, select: false },
});

export default instanceModel<User>('users', userSchema);
