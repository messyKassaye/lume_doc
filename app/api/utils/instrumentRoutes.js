/* eslint-disable max-statements */

import { validation } from 'api/utils';

const createSpy = (key, resolve) =>
  jasmine.createSpy(key).and.callFake((...args) => {
    resolve(`${key}:${args.join(',')}`);
  });

const mockResponseAsWritableStream = (res, resolve) => {
  let response = '';
  res.on = () => {};
  res.once = () => {};
  res.emit = () => {};
  res.write = data => {
    response += data.toString('utf8');
  };
  res.end = () => {
    resolve(response);
  };
};

const executeRoute = (method, routePath, req = {}, res, next = () => {}, app, runRoute = true) => {
  const args = app[method].calls.allArgs().find(a => a[0] === routePath);
  if (!args) {
    throw new Error(`Route ${method.toUpperCase()} ${routePath} is not defined`);
  }
  const result = new Promise((resolve, reject) => {
    let statusCode;
    let resType;
    mockResponseAsWritableStream(res, resolve);

    res.status = code => {
      statusCode = code;
    };

    res.type = type => {
      resType = type;
    };

    res.download = createSpy('download', resolve);
    res.redirect = createSpy('redirect', resolve);
    res.sendFile = createSpy('sendFile', resolve);
    res.send = createSpy('send', resolve);

    res.json = jasmine.createSpy('json').and.callFake(response => {
      if (statusCode) {
        response.status = statusCode;
      }
      if (resType) {
        response.type = resType;
      }
      resolve(response);
    });

    if (!args[1]) {
      return reject(new Error('route function has not been defined !'));
    }

    if (!runRoute) {
      return resolve();
    }

    const mockedNext = error => {
      next(error);
      reject(error);
    };

    return args[args.length - 1](req, res, mockedNext);
  });

  if (args) {
    result.middlewares = args.slice(1, -1).filter(arg => !arg.isJoi);
    const joiValidation = args.slice(1, -1).find(arg => arg.isJoi);
    if (joiValidation) {
      result.validation = joiValidation.describe();
    }
  }

  return result;
};

export default (route, io) => {
  const app = jasmine.createSpyObj('app', ['get', 'post', 'delete', 'use', 'options']);
  const originalValidateRequest = validation.validateRequest;
  spyOn(validation, 'validateRequest').and.callFake(schema => schema);
  route(app, io);
  validation.validateRequest = originalValidateRequest;

  const get = (routePath, req, res = {}, next) =>
    executeRoute('get', routePath, req, res, next, app);
  const _get = (routePath, req, res = {}, next) =>
    executeRoute('get', routePath, req, res, next, app, false);
  get.validation = (routePath, req, res = {}, next) =>
    executeRoute('get', routePath, req, res, next, app, false).validation;

  const post = (routePath, req, res = {}, next) =>
    executeRoute('post', routePath, req, res, next, app);
  const _post = (routePath, req, res = {}, next) =>
    executeRoute('post', routePath, req, res, next, app, false);
  post.validation = (routePath, req, res = {}, next) =>
    executeRoute('post', routePath, req, res, next, app, false).validation;

  const remove = (routePath, req, res = {}, next) =>
    executeRoute('delete', routePath, req, res, next, app);
  const _remove = (routePath, req, res = {}, next) =>
    executeRoute('delete', routePath, req, res, next, app, false);
  remove.validation = (routePath, req, res = {}, next) =>
    executeRoute('delete', routePath, req, res, next, app, false).validation;

  const instrumentedRoute = {
    get,
    _get,
    delete: remove,
    _delete: _remove,
    post,
    _post,
  };

  return instrumentedRoute;
};
