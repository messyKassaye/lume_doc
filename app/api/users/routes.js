/** @format */

import Joi from 'joi';

import { validation } from 'api/utils';
import needsAuthorization from '../auth/authMiddleware';
import users from './users';

const getDomain = req => `${req.protocol}://${req.get('host')}`;
export default app => {
  app.post(
    '/api/users',

    needsAuthorization(['admin', 'editor']),

    validation.validateRequest(
      Joi.object()
        .keys({
          _id: Joi.objectId().required(),
          __v: Joi.number(),
          username: Joi.string(),
          email: Joi.string(),
          password: Joi.string(),
          role: Joi.string().valid('admin', 'editor'),
          using2fa: Joi.boolean(),
        })
        .required()
    ),

    (req, res, next) => {
      users
        .save(req.body, req.user, getDomain(req))
        .then(response => res.json(response))
        .catch(next);
    }
  );

  app.post(
    '/api/users/new',
    needsAuthorization(),
    validation.validateRequest(
      Joi.object()
        .keys({
          username: Joi.string().required(),
          email: Joi.string().required(),
          password: Joi.string(),
          role: Joi.string()
            .valid('admin', 'editor')
            .required(),
        })
        .required()
    ),
    (req, res, next) => {
      users
        .newUser(req.body, getDomain(req))
        .then(response => res.json(response))
        .catch(next);
    }
  );

  app.post(
    '/api/unlockaccount',
    validation.validateRequest(
      Joi.object()
        .keys({
          username: Joi.string().required(),
          code: Joi.string().required(),
        })
        .required()
    ),
    (req, res, next) => {
      users
        .unlockAccount(req.body)
        .then(() => res.json('OK'))
        .catch(next);
    }
  );

  app.post(
    '/api/recoverpassword',
    validation.validateRequest(
      Joi.object()
        .keys({
          email: Joi.string().required(),
        })
        .required()
    ),
    (req, res, next) => {
      users
        .recoverPassword(req.body.email, getDomain(req))
        .then(() => res.json('OK'))
        .catch(next);
    }
  );

  app.post(
    '/api/resetpassword',
    validation.validateRequest(
      Joi.object()
        .keys({
          key: Joi.string().required(),
          password: Joi.string().required(),
        })
        .required()
    ),
    (req, res, next) => {
      users
        .resetPassword(req.body)
        .then(response => res.json(response))
        .catch(next);
    }
  );

  app.get('/api/users', needsAuthorization(), (_req, res, next) => {
    users
      .get()
      .then(response => res.json(response))
      .catch(next);
  });

  app.delete(
    '/api/users',
    needsAuthorization(),
    validation.validateRequest(
      Joi.object()
        .keys({
          _id: Joi.string().required(),
        })
        .required(),
      'query'
    ),
    (req, res, next) => {
      users
        .delete(req.query._id, req.user)
        .then(response => res.json(response))
        .catch(next);
    }
  );
};
