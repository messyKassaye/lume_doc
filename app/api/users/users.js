import SHA256 from 'crypto-js/sha256';
import crypto from 'crypto';

import { createError } from 'api/utils';
import random from 'shared/uniqueID';
import encryptPassword, { comparePasswords } from 'api/auth/encryptPassword';
import * as usersUtils from 'api/auth2fa/usersUtils';

import mailer from '../utils/mailer';
import model from './usersModel';
import passwordRecoveriesModel from './passwordRecoveriesModel';
import settings from '../settings/settings';

const MAX_FAILED_LOGIN_ATTEMPTS = 6;

const generateUnlockCode = () => crypto.randomBytes(32).toString('hex');

function conformRecoverText(options, _settings, domain, key, user) {
  const response = {};
  if (!options.newUser) {
    response.subject = 'Password set';
    response.text = `To set your password click on the following link:\n${domain}/setpassword/${key}`;
  }

  if (options.newUser) {
    const siteName = _settings.site_name || 'Uwazi';
    const text =
      'Hello!\n\n' +
      `The administrators of ${siteName} have created an account for you under the user name:\n` +
      `${user.username}\n\n` +
      'To complete this process, please create a strong password by clicking on the following link:\n' +
      `${domain}/setpassword/${key}?createAccount=true\n\n` +
      'For more information about the Uwazi platform, visit https://www.uwazi.io.\n\nThank you!\nUwazi team';

    const htmlLink = `<a href="${domain}/setpassword/${key}?createAccount=true">${domain}/setpassword/${key}?createAccount=true</a>`;

    response.subject = `Welcome to ${siteName}`;
    response.text = text;
    response.html = `<p>${response.text
      .replace(new RegExp(user.username, 'g'), `<b>${user.username}</b>`)
      .replace(new RegExp(`${domain}/setpassword/${key}\\?createAccount=true`, 'g'), htmlLink)
      .replace(
        new RegExp('https://www.uwazi.io', 'g'),
        '<a href="https://www.uwazi.io">https://www.uwazi.io</a>'
      )
      .replace(/\n{2,}/g, '</p><p>')
      .replace(/\n/g, '<br />')}</p>`;
  }
  return response;
}

const sendAccountLockedEmail = async (user, domain) => {
  const url = `${domain}/unlockaccount/${user.username}/${user.accountUnlockCode}`;
  const htmlLink = `<a href="${url}">${url}</a>`;
  const text =
    'Hello,\n\n' +
    'Your account has been locked because of too many failed login attempts. ' +
    'To unlock your account open the following link:\n' +
    `${url}`;
  const html = `<p>${text.replace(url, htmlLink)}</p>`;

  const settingsDetails = await settings.get();
  const emailSender = mailer.createSenderDetails(settingsDetails);
  const mailOptions = {
    from: emailSender,
    to: user.email,
    subject: 'Account locked',
    text,
    html,
  };

  return mailer.send(mailOptions);
};

const validateUserStatus = user => {
  if (!user) {
    throw createError('Invalid username or password', 401);
  }
  if (user.accountLocked) {
    throw createError('Account locked. Check your email to unlock.', 403);
  }
};

const updateOldPassword = async (user, password) => {
  await model.save({ _id: user._id, password: await encryptPassword(password) });
};

const blockAccount = async (user, domain) => {
  const accountUnlockCode = generateUnlockCode();
  const lockedUser = await model.db.findOneAndUpdate(
    { _id: user._id },
    { $set: { accountLocked: true, accountUnlockCode } },
    { new: true, fields: '+accountUnlockCode' }
  );
  await sendAccountLockedEmail(lockedUser, domain);
};

const newFailedLogin = async (user, domain) => {
  const updatedUser = await model.db.findOneAndUpdate(
    { _id: user._id },
    { $inc: { failedLogins: 1 } },
    { new: true, fields: '+failedLogins' }
  );
  if (updatedUser.failedLogins >= MAX_FAILED_LOGIN_ATTEMPTS) {
    await blockAccount(user, domain);
    throw createError('Account locked. Check your email to unlock.', 403);
  }
};

const validateUserPassword = async (user, password, domain) => {
  const passwordValidated = await comparePasswords(password, user.password);
  const oldPasswordValidated = user.password === SHA256(password).toString();

  if (oldPasswordValidated) {
    await updateOldPassword(user, password);
  }

  if (!oldPasswordValidated && !passwordValidated) {
    await newFailedLogin(user, domain);
    throw createError('Invalid username or password', 401);
  }
};

const validate2fa = async (user, token, domain) => {
  if (user.using2fa) {
    if (!token) {
      throw createError('Two-step verification token required', 409);
    }

    try {
      await usersUtils.verifyToken(user, token);
    } catch (err) {
      await newFailedLogin(user, domain);
      throw err;
    }
  }
};

const sanitizeUser = user => {
  const { password, accountLocked, failedLogins, accountUnlockCode, ...sanitizedUser } = user;
  return sanitizedUser;
};

export default {
  async save(user, currentUser) {
    const [userInTheDatabase] = await model.get({ _id: user._id }, '+password');

    if (user._id === currentUser._id.toString() && user.role !== currentUser.role) {
      return Promise.reject(createError('Can not change your own role', 403));
    }

    if (
      user.hasOwnProperty('role') &&
      user.role !== userInTheDatabase.role &&
      currentUser.role !== 'admin'
    ) {
      return Promise.reject(createError('Unauthorized', 403));
    }

    const { using2fa, secret, ...userToSave } = user;

    return model.save({
      ...userToSave,
      password: user.password ? await encryptPassword(user.password) : userInTheDatabase.password,
    });
  },

  async newUser(user, domain) {
    const [userNameMatch, emailMatch] = await Promise.all([
      model.get({ username: user.username }),
      model.get({ email: user.email }),
    ]);
    if (userNameMatch.length) {
      return Promise.reject(createError('Username already exists', 409));
    }
    if (emailMatch.length) {
      return Promise.reject(createError('Email already exists', 409));
    }
    const password = user.password ? user.password : random();
    const _user = await model.save({
      ...user,
      password: await encryptPassword(password),
      using2fa: undefined,
      secret: undefined,
    });
    await this.recoverPassword(user.email, domain, { newUser: true });
    return _user;
  },

  get(query, select) {
    return model.get(query, select);
  },

  getById(id, select = '') {
    return model.getById(id, select);
  },

  delete(_id, currentUser) {
    if (_id === currentUser._id.toString()) {
      return Promise.reject(createError('Can not delete yourself', 403));
    }

    return model.count().then(count => {
      if (count > 1) {
        return model.delete({ _id });
      }

      return Promise.reject(createError('Can not delete last user', 403));
    });
  },

  async login({ username, password, token }, domain) {
    const [user] = await this.get(
      { username },
      '+password +accountLocked +failedLogins +accountUnlockCode'
    );

    validateUserStatus(user);
    await validateUserPassword(user, password, domain);
    await validate2fa(user, token, domain);
    await model.db.updateOne({ _id: user._id }, { $unset: { failedLogins: 1 } });

    return sanitizeUser(user);
  },

  async unlockAccount({ username, code }) {
    const [user] = await model.get({ username, accountUnlockCode: code }, '_id');

    if (!user) {
      throw createError('Invalid username or unlock code', 403);
    }

    return model.save({
      ...user,
      accountLocked: false,
      accountUnlockCode: false,
      failedLogins: false,
    });
  },

  recoverPassword(email, domain, options = {}) {
    const key = SHA256(email + Date.now()).toString();
    return Promise.all([model.get({ email }), settings.get()]).then(([_user, _settings]) => {
      const user = _user[0];
      if (user) {
        return passwordRecoveriesModel.save({ key, user: user._id }).then(() => {
          const emailSender = mailer.createSenderDetails(_settings);
          const mailOptions = { from: emailSender, to: email };
          const mailTexts = conformRecoverText(options, _settings, domain, key, user);
          mailOptions.subject = mailTexts.subject;
          mailOptions.text = mailTexts.text;

          if (options.newUser) {
            mailOptions.html = mailTexts.html;
          }

          return mailer.send(mailOptions);
        });
      }

      return Promise.reject(createError('User not found', 403));
    });
  },

  async resetPassword(credentials) {
    const [key] = await passwordRecoveriesModel.get({ key: credentials.key });
    if (key) {
      return Promise.all([
        passwordRecoveriesModel.delete(key._id),
        model.save({ _id: key.user, password: await encryptPassword(credentials.password) }),
      ]);
    }
    throw createError('key not found', 403);
  },
};
