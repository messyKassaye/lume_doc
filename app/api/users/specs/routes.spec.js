import { catchErrors } from 'api/utils/jasmineHelpers';
import instrumentRoutes from 'api/utils/instrumentRoutes';

import userRoutes from '../routes.js';
import users from '../users.js';

describe('users routes', () => {
  let routes;

  beforeEach(() => {
    routes = instrumentRoutes(userRoutes);
  });

  describe('POST', () => {
    describe('/users', () => {
      it('should have a validation schema', () => {
        expect(routes.post.validation('/api/users')).toMatchSnapshot();
      });

      it('should call users save with the body', done => {
        spyOn(users, 'save').and.returnValue(Promise.resolve());
        const req = {
          body: 'changes',
          user: { _id: 'currentUser' },
          protocol: 'http',
          get: () => 'localhost',
        };
        routes
          .post('/api/users', req)
          .then(() => {
            expect(users.save).toHaveBeenCalledWith(
              'changes',
              { _id: 'currentUser' },
              'http://localhost'
            );
            done();
          })
          .catch(catchErrors(done));
      });
    });

    describe('/users/new', () => {
      it('should have a validation schema', () => {
        expect(routes.post.validation('/api/users/new')).toMatchSnapshot();
      });

      it('should call users newUser with the body', done => {
        spyOn(users, 'newUser').and.returnValue(Promise.resolve());
        const req = {
          body: 'changes',
          user: { _id: 'currentUser' },
          protocol: 'http',
          get: () => 'localhost',
        };
        routes
          .post('/api/users/new', req)
          .then(() => {
            expect(users.newUser).toHaveBeenCalledWith('changes', 'http://localhost');
            done();
          })
          .catch(catchErrors(done));
      });
    });

    describe('/recoverpassword', () => {
      it('should have a validation schema', () => {
        expect(routes.post.validation('/api/recoverpassword')).toMatchSnapshot();
      });

      it('should call users update with the body email', done => {
        spyOn(users, 'recoverPassword').and.returnValue(Promise.resolve());
        const req = { body: { email: 'recover@me.com' }, protocol: 'http', get: () => 'localhost' };
        routes
          .post('/api/recoverpassword', req)
          .then(response => {
            expect(response).toBe('OK');
            expect(users.recoverPassword).toHaveBeenCalledWith(
              'recover@me.com',
              'http://localhost'
            );
            done();
          })
          .catch(catchErrors(done));
      });

      it('should return an error if recover password fails', async () => {
        spyOn(users, 'recoverPassword').and.returnValue(Promise.reject(new Error('error')));
        const req = { body: { email: 'recover@me.com' }, protocol: 'http', get: () => 'localhost' };

        try {
          await routes.post('/api/recoverpassword', req);
        } catch (error) {
          expect(error).toEqual(new Error('error'));
        }
      });
    });

    describe('/resetpassword', () => {
      it('should have a validation schema', () => {
        expect(routes.post.validation('/api/resetpassword')).toMatchSnapshot();
      });

      it('should call users update with the body', done => {
        spyOn(users, 'resetPassword').and.returnValue(Promise.resolve());
        const req = { body: 'changes' };
        routes
          .post('/api/resetpassword', req)
          .then(() => {
            expect(users.resetPassword).toHaveBeenCalledWith('changes');
            done();
          })
          .catch(catchErrors(done));
      });
    });

    describe('/unlockaccount', () => {
      it('should have a validation schema', () => {
        expect(routes.post.validation('/api/unlockaccount')).toMatchSnapshot();
      });

      it('should call users.unlockAccount with the body', done => {
        jest.spyOn(users, 'unlockAccount').mockResolvedValue();
        const req = { body: 'credentials' };
        routes
          .post('/api/unlockaccount', req)
          .then(() => {
            expect(users.unlockAccount).toHaveBeenCalledWith('credentials');
            done();
          })
          .catch(catchErrors(done));
      });
    });
  });

  describe('GET', () => {
    it('should need authorization', () => {
      spyOn(users, 'get').and.returnValue(Promise.resolve(['users']));
      const req = {};
      expect(routes.get('/api/users', req)).toNeedAuthorization();
    });

    it('should call users get', done => {
      spyOn(users, 'get').and.returnValue(Promise.resolve(['users']));
      const req = {};
      routes
        .get('/api/users', req)
        .then(res => {
          expect(users.get).toHaveBeenCalled();
          expect(res).toEqual(['users']);
          done();
        })
        .catch(catchErrors(done));
    });

    it('should call next on error', async () => {
      spyOn(users, 'recoverPassword').and.returnValue(Promise.reject(new Error('error')));
      const req = { body: { email: 'recover@me.com' }, protocol: 'http', get: () => 'localhost' };

      try {
        await routes.post('/api/recoverpassword', req);
      } catch (error) {
        expect(error).toEqual(new Error('error'));
      }
    });
  });

  describe('DELETE', () => {
    let req;
    beforeEach(() => {
      req = { query: { _id: 123, username: 'Nooooooo!' }, user: { _id: 'currentUser' } };
      spyOn(users, 'delete').and.returnValue(Promise.resolve({ json: 'ok' }));
    });

    it('should have a validation schema', () => {
      expect(routes.delete.validation('/api/users')).toMatchSnapshot();
    });

    it('should need authorization', () => {
      expect(routes.delete('/api/users', req)).toNeedAuthorization();
    });

    it('should use users to delete it', done => {
      routes
        .delete('/api/users', req)
        .then(() => {
          expect(users.delete).toHaveBeenCalledWith(req.query._id, { _id: 'currentUser' });
          done();
        })
        .catch(catchErrors(done));
    });
  });
});
