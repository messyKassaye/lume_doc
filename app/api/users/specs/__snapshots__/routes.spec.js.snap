// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`users routes DELETE should have a validation schema 1`] = `
Object {
  "children": Object {
    "_id": Object {
      "flags": Object {
        "presence": "required",
      },
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
  },
  "flags": Object {
    "presence": "required",
  },
  "type": "object",
}
`;

exports[`users routes POST /recoverpassword should have a validation schema 1`] = `
Object {
  "children": Object {
    "email": Object {
      "flags": Object {
        "presence": "required",
      },
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
  },
  "flags": Object {
    "presence": "required",
  },
  "type": "object",
}
`;

exports[`users routes POST /resetpassword should have a validation schema 1`] = `
Object {
  "children": Object {
    "key": Object {
      "flags": Object {
        "presence": "required",
      },
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
    "password": Object {
      "flags": Object {
        "presence": "required",
      },
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
  },
  "flags": Object {
    "presence": "required",
  },
  "type": "object",
}
`;

exports[`users routes POST /unlockaccount should have a validation schema 1`] = `
Object {
  "children": Object {
    "code": Object {
      "flags": Object {
        "presence": "required",
      },
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
    "username": Object {
      "flags": Object {
        "presence": "required",
      },
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
  },
  "flags": Object {
    "presence": "required",
  },
  "type": "object",
}
`;

exports[`users routes POST /users should have a validation schema 1`] = `
Object {
  "children": Object {
    "__v": Object {
      "invalids": Array [
        Infinity,
        -Infinity,
      ],
      "type": "number",
    },
    "_id": Object {
      "flags": Object {
        "presence": "required",
      },
      "invalids": Array [
        "",
      ],
      "rules": Array [
        Object {
          "arg": Object {
            "pattern": /\\^\\[0-9a-fA-F\\]\\{24\\}\\$/,
          },
          "name": "regex",
        },
      ],
      "type": "string",
    },
    "email": Object {
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
    "password": Object {
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
    "role": Object {
      "flags": Object {
        "allowOnly": true,
      },
      "invalids": Array [
        "",
      ],
      "type": "string",
      "valids": Array [
        "admin",
        "editor",
      ],
    },
    "username": Object {
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
    "using2fa": Object {
      "falsy": Array [
        false,
      ],
      "flags": Object {
        "insensitive": true,
      },
      "truthy": Array [
        true,
      ],
      "type": "boolean",
    },
  },
  "flags": Object {
    "presence": "required",
  },
  "type": "object",
}
`;

exports[`users routes POST /users/new should have a validation schema 1`] = `
Object {
  "children": Object {
    "email": Object {
      "flags": Object {
        "presence": "required",
      },
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
    "password": Object {
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
    "role": Object {
      "flags": Object {
        "allowOnly": true,
        "presence": "required",
      },
      "invalids": Array [
        "",
      ],
      "type": "string",
      "valids": Array [
        "admin",
        "editor",
      ],
    },
    "username": Object {
      "flags": Object {
        "presence": "required",
      },
      "invalids": Array [
        "",
      ],
      "type": "string",
    },
  },
  "flags": Object {
    "presence": "required",
  },
  "type": "object",
}
`;
